beluga/include/beluga/algorithm/estimation.hpp

```diff
@@ -15,388 +15,502 @@
-      });
-
-  const auto number_of_non_zero_weights =
-      static_cast<Scalar>(ranges::count_if(weights_view, [&](auto weight) { return weight > 0; }));
-
-  const Scalar weighted_variance =
-      weighted_squared_deviations * number_of_non_zero_weights / (number_of_non_zero_weights - 1);
-
-  return std::pair{weighted_mean, weighted_variance};
-}
-
-/// Returns a pair consisting of the estimated mean pose and its covariance.
-/**
- * Given a range of poses, computes the estimated pose by averaging the translation
- * and rotation parts.
- * Computes the covariance matrix of the translation and rotation parts to create a 6x6 covariance matrix, assuming all
- * the poses are equally weighted.
+ * It supports floating-point numbers, vectors, and Lie group elements like SE2 and SE3.
  *
- * \tparam Poses A [sized range](https://en.cppreference.com/w/cpp/ranges/sized_range) type whose
- *  value type is `Sophus::SE3<Scalar>`.
- * \tparam Weights A [sized range](https://en.cppreference.com/w/cpp/ranges/sized_range) type whose
- *  value type is `Scalar`.
- * \tparam Pose The pose value type of the given range.
- * \tparam Scalar A scalar type, e.g. double or float.
- * \param poses 3D poses to estimate mean and covariances from, equally.
- * \return The estimated pose and its 6x6 covariance matrix.
+ * For Lie group elements, the function computes covariance in the tangent space, representing the uncertainty of poses
+ * on the Lie manifold. Users may perform the appropriate conversions to get the covariance matrix into their
+ * parametrization of interest.
+ *
+ * The weights are assumed to already be normalized. Non-normalized weights will yield incorrect results.
  */
-template <
-    class Poses,
-    class Pose = ranges::range_value_t<Poses>,
-    class Scalar = typename Pose::Scalar,
-    typename = std::enable_if_t<std::is_same_v<Pose, typename Sophus::SE3<Scalar>>>>
-std::pair<Sophus::SE3<Scalar>, Sophus::Matrix6<Scalar>> estimate(const Poses& poses) {
-  return beluga::estimate(poses, ranges::views::repeat_n(1.0, static_cast<std::ptrdiff_t>(poses.size())));
-}
-
-/// Returns a pair consisting of the estimated mean pose and its covariance.
+inline constexpr detail::covariance_fn covariance;
+
+namespace detail {
+
+/// \cond detail
+
+struct estimate_fn {
+  template <
+      class Values,
+      class Weights,
+      class Value = std::decay_t<ranges::range_value_t<Values>>,
+      class = std::enable_if_t<std::disjunction_v<
+          std::is_floating_point<Value>,
+          std::is_base_of<Eigen::MatrixBase<Value>, Value>,
+          std::is_base_of<Sophus::SE3Base<Value>, Value>>>>
+  auto operator()(Values&& values, Weights&& weights) const {
+    static_assert(ranges::forward_range<Values>);   // must allow multi-pass
+    static_assert(ranges::forward_range<Weights>);  // must allow multi-pass
+
+    auto normalized_weights = weights | ranges::views::transform([sum = ranges::accumulate(weights, 0.0)](auto weight) {
```

Comment:
@nahueespinosa shall we materialize the view in a vector? or are we banking on the compiler to inline and not recompute this twice? Same elsewhere.

Reply:
(1) Done in https://github.com/Ekumen-OS/beluga/pull/445/commits/434cbf49992b962b1d3c529dc435437a9ca74293.